### ISSUE TITLE
pylint complaint about missing dataclass constructor parameters with new astroid 2.12.6+

### ISSUE THREAD

**Issue by altendky** [PyCQA/pylint#7425]
### Bug description

We have some code using dataclasses, union hints, inheritance, and default factories where pylint started complaining about the parameter being missing when constructing an instance despite the default factory.  This started with astroid 2.12.6 and 2.12.7. 
 With astroid 2.12.5, pylint does not complain.  I've narrowed it down to being triggered by the inheritance and the union hint on the parent class, seemingly.

```python
"""a docstring for pylint"""
from dataclasses import dataclass
from typing import Union

@dataclass
class BadExampleParentClass:
    """a docstring for pylint"""
    xyz: Union[str, int]

@dataclass
class BadExampleClass(BadExampleParentClass):
    """a docstring for pylint"""
    xyz: str = ""

bad_c = BadExampleClass()

@dataclass
class GoodExampleParentClass:
    """a docstring for pylint"""
    xyz: str

@dataclass
class GoodExampleClass(GoodExampleParentClass):
    """a docstring for pylint"""
    xyz: str = ""

good_c = GoodExampleClass()
```

```console
$ python3.10 -m venv venv
```

```console
$ venv/bin/python -m pip install --upgrade pip setuptools wheel
Requirement already satisfied: pip in ./venv/lib/python3.10/site-packages (21.2.4)
Collecting pip
  Using cached pip-22.2.2-py3-none-any.whl (2.0 MB)
Requirement already satisfied: setuptools in ./venv/lib/python3.10/site-packages (58.1.0)
Collecting setuptools
  Using cached setuptools-65.3.0-py3-none-any.whl (1.2 MB)
Collecting wheel
  Using cached wheel-0.37.1-py2.py3-none-any.whl (35 kB)
Installing collected packages: wheel, setuptools, pip
  Attempting uninstall: setuptools
    Found existing installation: setuptools 58.1.0
    Uninstalling setuptools-58.1.0:
      Successfully uninstalled setuptools-58.1.0
  Attempting uninstall: pip
    Found existing installation: pip 21.2.4
    Uninstalling pip-21.2.4:
      Successfully uninstalled pip-21.2.4
Successfully installed pip-22.2.2 setuptools-65.3.0 wheel-0.37.1
```

```console
$ venv/bin/python -m pip install pylint==2.15.0 astroid==2.12.7
Collecting pylint==2.15.0
  Using cached pylint-2.15.0-py3-none-any.whl (505 kB)
Collecting astroid==2.12.7
  Using cached astroid-2.12.7-py3-none-any.whl (262 kB)
Collecting mccabe<0.8,>=0.6
  Using cached mccabe-0.7.0-py2.py3-none-any.whl (7.3 kB)
Collecting isort<6,>=4.2.5
  Using cached isort-5.10.1-py3-none-any.whl (103 kB)
Collecting tomlkit>=0.10.1
  Using cached tomlkit-0.11.4-py3-none-any.whl (35 kB)
Collecting platformdirs>=2.2.0
  Using cached platformdirs-2.5.2-py3-none-any.whl (14 kB)
Collecting dill>=0.2
  Using cached dill-0.3.5.1-py2.py3-none-any.whl (95 kB)
Collecting tomli>=1.1.0
  Using cached tomli-2.0.1-py3-none-any.whl (12 kB)
Collecting wrapt<2,>=1.11
  Using cached wrapt-1.14.1-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (77 kB)
Collecting lazy-object-proxy>=1.4.0
  Using cached lazy_object_proxy-1.7.1-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (62 kB)
Installing collected packages: wrapt, tomlkit, tomli, platformdirs, mccabe, lazy-object-proxy, isort, dill, astroid, pylint
Successfully installed astroid-2.12.7 dill-0.3.5.1 isort-5.10.1 lazy-object-proxy-1.7.1 mccabe-0.7.0 platformdirs-2.5.2 pylint-2.15.0 tomli-2.0.1 tomlkit-0.11.4 wrapt-1.14.1
```

```console
$ cat > example.py << EOF
> """a docstring for pylint"""
> from dataclasses import dataclass
> from typing import Union
> 
> @dataclass
> class BadExampleParentClass:
>     """a docstring for pylint"""
>     xyz: Union[str, int]
> 
> @dataclass
> class BadExampleClass(BadExampleParentClass):
>     """a docstring for pylint"""
>     xyz: str = ""
> 
> bad_c = BadExampleClass()
> 
> @dataclass
> class GoodExampleParentClass:
>     """a docstring for pylint"""
>     xyz: str
> 
> @dataclass
> class GoodExampleClass(GoodExampleParentClass):
>     """a docstring for pylint"""
>     xyz: str = ""
> 
> good_c = GoodExampleClass()
> EOF
```

```console
$ venv/bin/pylint example.py
************* Module example
example.py:15:8: E1120: No value for argument 'xyz' in constructor call (no-value-for-parameter)

------------------------------------------------------------------
Your code has been rated at 5.83/10 (previous run: 5.83/10, +0.00)

```

### Command used

```shell
venv/bin/pylint example.py
```

### Pylint output

```shell
************* Module example
example.py:15:8: E1120: No value for argument 'xyz' in constructor call (no-value-for-parameter)

------------------------------------------------------------------
Your code has been rated at 5.83/10 (previous run: 5.83/10, +0.00)
```

### Expected behavior

No complain as happened with astroid 2.12.5.

### Pylint version

```shell
pylint 2.15.0
astroid 2.12.7
```

### OS / Environment

Ubuntu Linux 20.04

**Comment by altendky** [comment]
For reference, here are the astroid diffs.

https://github.com/PyCQA/astroid/compare/v2.12.5...v2.12.6
https://github.com/PyCQA/astroid/compare/v2.12.6...v2.12.7

**Comment by DanielNoord** [comment]
Thanks for the report. I pushed a PR with a fix to `astroid`. We're probably going to release another patch version of `astroid` and include that in `pylint` as well.

A recent PR for `dataclasses` has caused a lot of unexpected regression. Sorry about that!
